File: packages/json/llm-json-stream-ts/packages/llm-json-stream/src/classes/property_stream.ts
/**
 * Property stream classes providing reactive access to JSON values.
 * These are the public API classes that users interact with.
 *
 * Uses native async iterators as the primary streaming interface,
 * providing a modern, cross-platform approach that works in Node.js,
 * Deno, Bun, and browsers.
 */

import { JsonStreamParserController } from "./json_stream_parser.js";

/**
 * Internal interface for pushing values to the async iterator.
 * This is used by the stream controllers to emit values.
 */
export interface AsyncIteratorController<T> {
    push(value: T): void;
    complete(): void;
    error(err: Error): void;
}

/**
 * Creates an async iterator with a controller for pushing values.
 * This is the foundation for all streaming in the library.
 */
export function createAsyncIterator<T>(): {
    iterator: AsyncIterableIterator<T>;
    controller: AsyncIteratorController<T>;
} {
    const queue: T[] = [];
    let resolve: ((value: IteratorResult<T>) => void) | null = null;
    let isDone = false;
    let error: Error | null = null;

    const iterator: AsyncIterableIterator<T> = {
        [Symbol.asyncIterator]() {
            return this;
        },
        async next(): Promise<IteratorResult<T>> {
            // If there's an error, throw it
            if (error) {
                throw error;
            }

            // If there are queued values, return the next one
            if (queue.length > 0) {
                return { value: queue.shift()!, done: false };
            }

            // If done and queue is empty, signal completion
            if (isDone) {
                return { value: undefined as any, done: true };
            }

            // Wait for a value to be pushed
            return new Promise<IteratorResult<T>>((res) => {
                resolve = res;
            });
        },
    };

    const controller: AsyncIteratorController<T> = {
        push(value: T) {
            if (isDone) return;

            if (resolve) {
                resolve({ value, done: false });
                resolve = null;
            } else {
                queue.push(value);
            }
        },
        complete() {
            isDone = true;
            if (resolve) {
                resolve({ value: undefined as any, done: true });
                resolve = null;
            }
        },
        error(err: Error) {
            error = err;
            isDone = true;
            if (resolve) {
                // For errors, we need to reject the promise
                // We do this by creating a rejected result
                resolve({ value: undefined as any, done: true });
                resolve = null;
            }
        },
    };

    return { iterator, controller };
}

/**
 * Base class for all property streams.
 * Property streams provide access to JSON values as they are parsed from the input stream.
 *
 * The primary interface is async iteration:
 * ```typescript
 * for await (const chunk of propertyStream) {
 *   console.log(chunk);
 * }
 * ```
 */
export class PropertyStream<T> implements AsyncIterable<T> {
    protected _promise: Promise<T>;
    protected parserController: JsonStreamParserController;
    protected _iteratorController: AsyncIteratorController<T>;
    protected _iterator: AsyncIterableIterator<T>;

    // Buffered values for late subscribers
    protected _buffer: T[] = [];
    protected _isComplete = false;

    constructor(
        promise: Promise<T>,
        parserController: JsonStreamParserController,
        iteratorController?: AsyncIteratorController<T>,
        iterator?: AsyncIterableIterator<T>,
    ) {
        this._promise = promise;
        this.parserController = parserController;

        if (iteratorController && iterator) {
            this._iteratorController = iteratorController;
            this._iterator = iterator;
        } else {
            // Create default iterator
            const { iterator: iter, controller } = createAsyncIterator<T>();
            this._iterator = iter;
            this._iteratorController = controller;
        }
    }

    /**
     * A promise that resolves with the final parsed value of this property.
     * Use this when you need the complete value and don't need to react to partial chunks.
     */
    get promise(): Promise<T> {
        return this._promise;
    }

    /**
     * Implements the async iterable protocol, allowing consumption of the stream
     * using modern `for await...of` syntax.
     *
     * This is the BUFFERED stream - late subscribers will receive all previous values.
     *
     * @example
     * ```typescript
     * const nameStream = parser.getStringProperty("name");
     * for await (const chunk of nameStream) {
     *   console.log(chunk);
     * }
     * ```
     */
    [Symbol.asyncIterator](): AsyncIterableIterator<T> {
        return this._createBufferedIterator();
    }

    /**
     * Returns an unbuffered async iterator.
     * Late subscribers will NOT receive previous values - only new values from subscription point.
     *
     * @example
     * ```typescript
     * const nameStream = parser.getStringProperty("name");
     * for await (const chunk of nameStream.unbuffered()) {
     *   console.log(chunk); // Only new chunks from this point
     * }
     * ```
     */
    unbuffered(): AsyncIterableIterator<T> {
        return this._createUnbufferedIterator();
    }

    /**
     * Internal method to push a value to the stream.
     * Called by property stream controllers.
     */
    _pushValue(value: T): void {
        this._buffer.push(value);
        this._iteratorController.push(value);
    }

    /**
     * Internal method to signal stream completion.
     * Called by property stream controllers.
     */
    _complete(): void {
        this._isComplete = true;
        this._iteratorController.complete();
    }

    /**
     * Internal method to signal an error.
     * Called by property stream controllers.
     */
    _error(err: Error): void {
        this._isComplete = true;
        this._iteratorController.error(err);
    }

    /**
     * Creates a buffered iterator that replays all previous values.
     */
    private _createBufferedIterator(): AsyncIterableIterator<T> {
        let bufferIndex = 0;
        const self = this;

        return {
            [Symbol.asyncIterator]() {
                return this;
            },
            async next(): Promise<IteratorResult<T>> {
                // First, replay buffered values
                if (bufferIndex < self._buffer.length) {
                    return { value: self._buffer[bufferIndex++]!, done: false };
                }

                // If complete and no more buffered values, we're done
                if (self._isComplete) {
                    return { value: undefined as any, done: true };
                }

                // Wait for new live values
                return new Promise<IteratorResult<T>>((resolve) => {
                    const checkBuffer = () => {
                        if (bufferIndex < self._buffer.length) {
                            resolve({
                                value: self._buffer[bufferIndex++]!,
                                done: false,
                            });
                            return true;
                        }
                        if (self._isComplete) {
                            resolve({ value: undefined as any, done: true });
                            return true;
                        }
                        return false;
                    };

                    // Check periodically for new values
                    const interval = setInterval(() => {
                        if (checkBuffer()) {
                            clearInterval(interval);
                        }
                    }, 1);

                    // Also check immediately
                    if (checkBuffer()) {
                        clearInterval(interval);
                    }
                });
            },
        };
    }

    /**
     * Creates an unbuffered iterator that only receives new values.
     */
    private _createUnbufferedIterator(): AsyncIterableIterator<T> {
        const self = this;
        let lastCheckedIndex = this._buffer.length; // Start from current position

        return {
            [Symbol.asyncIterator]() {
                return this;
            },
            async next(): Promise<IteratorResult<T>> {
                // Check for new values since subscription
                if (
                    self._isComplete && lastCheckedIndex >= self._buffer.length
                ) {
                    return { value: undefined as any, done: true };
                }

                if (lastCheckedIndex < self._buffer.length) {
                    return {
                        value: self._buffer[lastCheckedIndex++]!,
                        done: false,
                    };
                }

                // Wait for next value
                return new Promise<IteratorResult<T>>((resolve) => {
                    const checkForNew = () => {
                        if (self._buffer.length > lastCheckedIndex) {
                            const newValue = self._buffer[lastCheckedIndex++]!;
                            resolve({ value: newValue, done: false });
                            return true;
                        }
                        if (self._isComplete) {
                            resolve({ value: undefined as any, done: true });
                            return true;
                        }
                        return false;
                    };

                    const interval = setInterval(() => {
                        if (checkForNew()) {
                            clearInterval(interval);
                        }
                    }, 1);

                    if (checkForNew()) {
                        clearInterval(interval);
                    }
                });
            },
        };
    }
}

/**
 * A property stream specifically for string values.
 * Strings emit chunks as they are parsed, allowing progressive display.
 */
export class StringPropertyStream extends PropertyStream<string> {
    constructor(
        promise: Promise<string>,
        parserController: JsonStreamParserController,
        iteratorController?: AsyncIteratorController<string>,
        iterator?: AsyncIterableIterator<string>,
    ) {
        super(promise, parserController, iteratorController, iterator);
    }
}

/**
 * A property stream for JSON object values.
 * Provides onProperty callback for reacting to properties as they start parsing.
 */
export class ObjectPropertyStream extends PropertyStream<Record<string, any>> {
    private propertyPath: string;
    private _onPropertyCallbacks: Array<
        (property: PropertyStream<any>, key: string) => void
    > = [];

    constructor(
        promise: Promise<Record<string, any>>,
        parserController: JsonStreamParserController,
        propertyPath: string,
        iteratorController?: AsyncIteratorController<Record<string, any>>,
        iterator?: AsyncIterableIterator<Record<string, any>>,
    ) {
        super(promise, parserController, iteratorController, iterator);
        this.propertyPath = propertyPath;
    }

    /**
     * Registers a callback that fires when each property in the object starts parsing.
     * The callback receives the property stream for the value and the property key.
     *
     * @example
     * ```typescript
     * const objectStream = parser.getObjectProperty('user');
     * objectStream.onProperty((property, key) => {
     *   console.log(`Property ${key} discovered`);
     *   if (property instanceof StringPropertyStream) {
     *     for await (const chunk of property) {
     *       console.log(`${key} chunk: ${chunk}`);
     *     }
     *   }
     * });
     * ```
     */
    onProperty(
        callback: (property: PropertyStream<any>, key: string) => void,
    ): void {
        this._onPropertyCallbacks.push(callback);
    }

    /**
     * Internal method to notify callbacks about a new property.
     */
    _notifyProperty(property: PropertyStream<any>, key: string): void {
        for (const callback of this._onPropertyCallbacks) {
            callback(property, key);
        }
    }
}

/**
 * A property stream for JSON array values.
 * Provides onElement callback for reacting to elements as they start parsing.
 */
export class ArrayPropertyStream<T = any> extends PropertyStream<T[]> {
    private propertyPath: string;
    private _onElementCallbacks: Array<
        (propertyStream: PropertyStream<any>, index: number) => void
    > = [];

    constructor(
        promise: Promise<T[]>,
        parserController: JsonStreamParserController,
        propertyPath: string,
        iteratorController?: AsyncIteratorController<T[]>,
        iterator?: AsyncIterableIterator<T[]>,
    ) {
        super(promise, parserController, iteratorController, iterator);
        this.propertyPath = propertyPath;
    }

    /**
     * Registers a callback that fires when each array element starts parsing.
     * The callback receives the property stream for the new element and its index.
     *
     * @example
     * ```typescript
     * const arrayStream = parser.getArrayProperty('items');
     * arrayStream.onElement((element, index) => {
     *   console.log(`Element ${index} discovered`);
     *   for await (const snapshot of element) {
     *     console.log(`Element ${index}: ${snapshot}`);
     *   }
     * });
     * ```
     */
    onElement(
        callback: (propertyStream: PropertyStream<any>, index: number) => void,
    ): void {
        this._onElementCallbacks.push(callback);
    }

    /**
     * Internal method to notify callbacks about a new element.
     */
    _notifyElement(propertyStream: PropertyStream<any>, index: number): void {
        for (const callback of this._onElementCallbacks) {
            callback(propertyStream, index);
        }
    }
}

/**
 * A property stream for number values.
 * Numbers emit once when the complete value is parsed.
 */
export class NumberPropertyStream extends PropertyStream<number> {
    constructor(
        promise: Promise<number>,
        parserController: JsonStreamParserController,
        iteratorController?: AsyncIteratorController<number>,
        iterator?: AsyncIterableIterator<number>,
    ) {
        super(promise, parserController, iteratorController, iterator);
    }
}

/**
 * A property stream for boolean values.
 * Booleans emit once when the complete value is parsed.
 */
export class BooleanPropertyStream extends PropertyStream<boolean> {
    constructor(
        promise: Promise<boolean>,
        parserController: JsonStreamParserController,
        iteratorController?: AsyncIteratorController<boolean>,
        iterator?: AsyncIterableIterator<boolean>,
    ) {
        super(promise, parserController, iteratorController, iterator);
    }
}

/**
 * A property stream for null values.
 * Null emits once when parsed.
 */
export class NullPropertyStream extends PropertyStream<null> {
    constructor(
        promise: Promise<null>,
        parserController: JsonStreamParserController,
        iteratorController?: AsyncIteratorController<null>,
        iterator?: AsyncIterableIterator<null>,
    ) {
        super(promise, parserController, iteratorController, iterator);
    }
}

File: packages/json/llm-json-stream-python/src/llm_json_stream/property_stream.py
"""Property streams exposing parsed JSON values."""

from __future__ import annotations

import asyncio
from typing import Any, AsyncIterator, Awaitable, Callable, Generic, List, Optional, Protocol, TypeVar

if False:  # pragma: no cover - for typing only
    from .json_stream_parser import JsonStreamParserController


T = TypeVar("T")


class _ParserControllerProtocol(Protocol):
    def get_property_stream(self, property_path: str, stream_type: str) -> "PropertyStream[Any]": ...


class PropertyStream(Generic[T]):
    """Base class for all property streams.

    Property streams are both AsyncIterables (for streaming chunks/snapshots)
    and Awaitables (for the final value).
    """

    _COMPLETE_SENTINEL = object()

    def __init__(
        self,
        future: asyncio.Future[T],
        parser_controller: _ParserControllerProtocol,
        property_path: str,
    ) -> None:
        self._future = future
        self._parser_controller = parser_controller
        self._property_path = property_path
        self._buffer: List[T] = []
        self._subscribers: List[asyncio.Queue[object]] = []
        self._is_complete = False
        self._error: Exception | None = None
        self._on_value_callbacks: List[Callable[[T], None]] = []

    @property
    def property_path(self) -> str:
        return self._property_path

    def __await__(self):  # type: ignore[override]
        return self._future.__await__()

    def __aiter__(self) -> AsyncIterator[T]:
        return self._create_iterator(buffered=True)

    def unbuffered(self) -> AsyncIterator[T]:
        """Return an iterator that only receives new values."""
        return self._create_iterator(buffered=False)

    def on_value(self, callback: Callable[[T], None]) -> None:
        """Register a callback for when the final value is available."""
        if self._future.done() and not self._future.cancelled():
            if self._future.exception() is None:
                callback(self._future.result())
            return
        self._on_value_callbacks.append(callback)

    def _push_value(self, value: T) -> None:
        if self._is_complete:
            return
        self._buffer.append(value)
        for queue in list(self._subscribers):
            queue.put_nowait(value)

    def _complete(self) -> None:
        if self._is_complete:
            return
        self._is_complete = True
        for queue in list(self._subscribers):
            queue.put_nowait(self._COMPLETE_SENTINEL)

    def _error_out(self, error: Exception) -> None:
        if self._is_complete:
            return
        self._error = error
        self._is_complete = True
        for queue in list(self._subscribers):
            queue.put_nowait(self._COMPLETE_SENTINEL)

    async def _iterator_loop(self, queue: asyncio.Queue[object]) -> AsyncIterator[T]:
        try:
            while True:
                item = await queue.get()
                if item is self._COMPLETE_SENTINEL:
                    if self._error is not None:
                        raise self._error
                    return
                yield item  # type: ignore[misc]
        finally:
            if queue in self._subscribers:
                self._subscribers.remove(queue)

    def _create_iterator(self, buffered: bool) -> AsyncIterator[T]:
        queue: asyncio.Queue[object] = asyncio.Queue()
        if buffered:
            for value in self._buffer:
                queue.put_nowait(value)
        self._subscribers.append(queue)
        if self._is_complete:
            queue.put_nowait(self._COMPLETE_SENTINEL)
        return self._iterator_loop(queue)

    def _notify_final_value(self, value: T) -> None:
        for callback in self._on_value_callbacks:
            callback(value)
        self._on_value_callbacks.clear()


class StringPropertyStream(PropertyStream[str]):
    pass


class NumberPropertyStream(PropertyStream[float]):
    pass


class BooleanPropertyStream(PropertyStream[bool]):
    pass


class NullPropertyStream(PropertyStream[None]):
    pass


class PropertyGetterMixin:
    """Mixin to expose path-based property access methods."""

    def build_property_path(self, key: str) -> str:  # pragma: no cover - interface
        raise NotImplementedError

    @property
    def parser_controller(self) -> _ParserControllerProtocol:  # pragma: no cover - interface
        raise NotImplementedError

    def get_string_property(self, key: str) -> StringPropertyStream:
        return self.parser_controller.get_property_stream(
            self.build_property_path(key), "string"
        )  # type: ignore[return-value]

    def get_number_property(self, key: str) -> NumberPropertyStream:
        return self.parser_controller.get_property_stream(
            self.build_property_path(key), "number"
        )  # type: ignore[return-value]

    def get_boolean_property(self, key: str) -> BooleanPropertyStream:
        return self.parser_controller.get_property_stream(
            self.build_property_path(key), "boolean"
        )  # type: ignore[return-value]

    def get_null_property(self, key: str) -> NullPropertyStream:
        return self.parser_controller.get_property_stream(
            self.build_property_path(key), "null"
        )  # type: ignore[return-value]

    def get_map_property(self, key: str) -> "MapPropertyStream":
        return self.parser_controller.get_property_stream(
            self.build_property_path(key), "object"
        )  # type: ignore[return-value]

    def get_list_property(self, key: str) -> "ListPropertyStream":
        return self.parser_controller.get_property_stream(
            self.build_property_path(key), "array"
        )  # type: ignore[return-value]

    def str(self, key: str) -> StringPropertyStream:
        return self.get_string_property(key)

    def number(self, key: str) -> NumberPropertyStream:
        return self.get_number_property(key)

    def boolean(self, key: str) -> BooleanPropertyStream:
        return self.get_boolean_property(key)

    def nil(self, key: str) -> NullPropertyStream:
        return self.get_null_property(key)

    def map(self, key: str) -> "MapPropertyStream":
        return self.get_map_property(key)

    def list(self, key: str) -> "ListPropertyStream":
        return self.get_list_property(key)


class MapPropertyStream(PropertyStream[dict[str, Any]], PropertyGetterMixin):
    def __init__(
        self,
        future: asyncio.Future[dict[str, Any]],
        parser_controller: _ParserControllerProtocol,
        property_path: str,
    ) -> None:
        super().__init__(future, parser_controller, property_path)
        self._on_property_callbacks: List[
            Callable[[PropertyStream[Any], str], None]
        ] = []

    def on_property(self, callback: Callable[[PropertyStream[Any], str], None]) -> None:
        self._on_property_callbacks.append(callback)

    def _notify_property(self, stream: PropertyStream[Any], key: str) -> None:
        for callback in self._on_property_callbacks:
            callback(stream, key)

    @property
    def parser_controller(self) -> _ParserControllerProtocol:
        return self._parser_controller

    def build_property_path(self, key: str) -> str:
        if not self._property_path:
            return key
        if key.startswith("["):
            return f"{self._property_path}{key}"
        return f"{self._property_path}.{key}"


class ListPropertyStream(PropertyStream[list[Any]], PropertyGetterMixin):
    def __init__(
        self,
        future: asyncio.Future[list[Any]],
        parser_controller: _ParserControllerProtocol,
        property_path: str,
    ) -> None:
        super().__init__(future, parser_controller, property_path)
        self._on_element_callbacks: List[
            Callable[[PropertyStream[Any], int], None]
        ] = []

    def on_element(
        self, callback: Callable[[PropertyStream[Any], int], None]
    ) -> None:
        self._on_element_callbacks.append(callback)

    def _notify_element(self, stream: PropertyStream[Any], index: int) -> None:
        for callback in self._on_element_callbacks:
            callback(stream, index)

    @property
    def parser_controller(self) -> _ParserControllerProtocol:
        return self._parser_controller

    def build_property_path(self, key: str) -> str:
        if not self._property_path:
            return key
        if key.startswith("["):
            return f"{self._property_path}{key}"
        return f"{self._property_path}.{key}"
